# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [4]:
# Load the dataset

import pandas as pd

url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
df_file1 = pd.read_csv(url)

url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv'
df_file2 = pd.read_csv(url)

url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv'
df_file3 = pd.read_csv(url)

In [8]:
# Combine the three datasets
pd.concat([df_file1, df_file2, df_file3], axis=0)

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount,State,Gender
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934,NaN,NaN
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935,NaN,NaN
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247,NaN,NaN
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344,NaN,NaN
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
7065,LA72316,NaN,NaN,Bachelor,23405.98798,71941.0,73.0,0,Personal Auto,Four-Door Car,198.234764,California,M
7066,PK87824,NaN,NaN,College,3096.511217,21604.0,79.0,0,Corporate Auto,Four-Door Car,379.200000,California,F
7067,TD14365,NaN,NaN,Bachelor,8163.890428,0.0,85.0,3,Corporate Auto,Four-Door Car,790.784983,California,M
7068,UP19263,NaN,NaN,College,7524.442436,21941.0,96.0,0,Personal Auto,Four-Door Car,691.200000,California,M


In [9]:
#New data frame combining the information from the 3 files
df_new = pd.concat([df_file1, df_file2, df_file3])
df_new

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount,State,Gender
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934,NaN,NaN
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935,NaN,NaN
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247,NaN,NaN
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344,NaN,NaN
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
7065,LA72316,NaN,NaN,Bachelor,23405.98798,71941.0,73.0,0,Personal Auto,Four-Door Car,198.234764,California,M
7066,PK87824,NaN,NaN,College,3096.511217,21604.0,79.0,0,Corporate Auto,Four-Door Car,379.200000,California,F
7067,TD14365,NaN,NaN,Bachelor,8163.890428,0.0,85.0,3,Corporate Auto,Four-Door Car,790.784983,California,M
7068,UP19263,NaN,NaN,College,7524.442436,21941.0,96.0,0,Personal Auto,Four-Door Car,691.200000,California,M


#Cleaning the data in the new dataframe

In [12]:
#st replaced for State
df_new.rename(columns= {'ST': 'State'}, inplace=True)
df_new.rename(columns= {'GENDER': 'Gender'}, inplace=True)
df_new.head(0)

,Customer,State,Gender,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount,State,Gender


In [ ]:
#Check duplicated columns
df_new.columns[df_new.columns.duplicated()]

Index(['State', 'Gender'], dtype='str')

In [ ]:
#Check first duplicated column
df_new["State"]

,State,State
0,Washington,NaN
1,Arizona,NaN
2,Nevada,NaN
3,California,NaN
4,Washington,NaN
...,...,...
7065,NaN,California
7066,NaN,California
7067,NaN,California
7068,NaN,California


In [15]:
df_new["State"].head(10)

,State,State
0,Washington,NaN
1,Arizona,NaN
2,Nevada,NaN
3,California,NaN
4,Washington,NaN
5,Oregon,NaN
6,Oregon,NaN
7,Arizona,NaN
8,Oregon,NaN
9,Oregon,NaN


In [20]:
#Antes de substituir as colunas originais, precisamos saber se existem linhas em que as duas State possuem valores ao mesmo tempo.
states = df_new["State"]

states.notna().sum(axis=1).value_counts()

#Isso significa que 9.137 linhas têm exatamente um valor de State entre as duas colunas, e 2.937 linhas não têm valor em nenhuma das duas. O mais importante: não apareceu 2, então não há linhas em que as duas colunas State estejam preenchidas ao mesmo tempo. Portanto, podemos combiná-las sem conflito.

1    9137
0    2937
Name: count, dtype: int64

In [ ]:
#Save new column State with the rows combined
state_combined = df_new["State"].bfill(axis=1).iloc[:, 0]

In [23]:
#Remove the 2 columns named State
df_new = df_new.drop(columns="State")

In [ ]:
#Add new combined column
df_new["State"] = state_combined
df_new["State"].head(10)

0    Washington
1       Arizona
2        Nevada
3    California
4    Washington
5        Oregon
6        Oregon
7       Arizona
8        Oregon
9        Oregon
Name: State, dtype: str

In [ ]:
#Check remaining duplicated columns
df_new.columns[df_new.columns.duplicated()]

Index(['Gender'], dtype='str')

In [ ]:
#Check columns Gender before consolidate 
genders = df_new["Gender"]

genders.notna().sum(axis=1).value_counts()

1    9015
0    3059
Name: count, dtype: int64

In [ ]:
#Save new column Gender with the rows combined
gender_combined = df_new["Gender"].bfill(axis=1).iloc[:, 0]

In [28]:
#Remove the 2 columns named Gender
df_new = df_new.drop(columns="Gender")

In [29]:
#Add new combined column
df_new["Gender"] = gender_combined

In [30]:
df_new.shape

(12074, 11)

In [ ]:
#Check all columns
df_new.head()

,Customer,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount,State,Gender
0,RB50392,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934,Washington,NaN
1,QZ44356,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935,Arizona,F
2,AI49188,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247,Nevada,F
3,WW63253,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344,California,M
4,GA49547,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323,Washington,M


In [33]:
#checking STATE column values
df_new["State"].unique()

<StringArray>
['Washington',    'Arizona',     'Nevada', 'California',     'Oregon',
       'Cali',         'AZ',         'WA',          nan]
Length: 9, dtype: str

In [34]:
#State mapping dictionary
state_mapping = {
    "AZ": "Arizona",
    "Cali": "California",
    "WA": "Washington"
}

In [36]:
#Replacing the values in the STATE column
df_new["State"] = df_new["State"].replace(state_mapping)
df_new["State"].unique()

<StringArray>
['Washington', 'Arizona', 'Nevada', 'California', 'Oregon', nan]
Length: 6, dtype: str

In [38]:
#Checkind Education column values
df_new["Education"].unique()

<StringArray>
[              'Master',             'Bachelor', 'High School or Below',
              'College',            'Bachelors',               'Doctor',
                    nan]
Length: 7, dtype: str

In [41]:
df_new["Education"] = df_new["Education"] .replace({"Bachelors": "Bachelor"})
df_new["Education"].unique()

<StringArray>
['Master', 'Bachelor', 'High School or Below', 'College', 'Doctor', nan]
Length: 6, dtype: str

In [42]:
#checking Customer Lifetime Value
df_new["Customer Lifetime Value"].head()

0            NaN
1     697953.59%
2    1288743.17%
3     764586.18%
4     536307.65%
Name: Customer Lifetime Value, dtype: object

In [45]:
#vehicle class
df_new["Vehicle Class"].unique()

<StringArray>
['Four-Door Car',  'Two-Door Car',           'SUV',    'Luxury SUV',
    'Sports Car',    'Luxury Car',             nan]
Length: 7, dtype: str

In [46]:
# vehicle class: Sports Car", "Luxury SUV" and "Luxury Car" replaced by "Luxury"
df_new["Vehicle Class"] = df_new["Vehicle Class"].replace({"Sports Car": "Luxury", "Luxury SUV": "Luxury", "Luxury Car": "Luxury"})
df_new["Vehicle Class"].unique()

<StringArray>
['Four-Door Car', 'Two-Door Car', 'SUV', 'Luxury', nan]
Length: 5, dtype: str

In [47]:
#Checking GENDER column values
df_new["Gender"].value_counts()

Gender
F         4560
M         4368
Male        40
female      30
Femal       17
Name: count, dtype: int64

In [48]:
# Fixing Gender column 
df_new["Gender"] = df_new["Gender"].replace({"Male": "M", "Female": "F", "Femal": "F", "female": "F" , "Female": "F"})

df_new["Gender"].value_counts()

Gender
F    4607
M    4408
Name: count, dtype: int64

In [49]:
df_final = df_new.copy()
df_final.shape
df_final.head()

,Customer,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount,State,Gender
0,RB50392,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934,Washington,NaN
1,QZ44356,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935,Arizona,F
2,AI49188,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247,Nevada,F
3,WW63253,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344,California,M
4,GA49547,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323,Washington,M


In [50]:
df_final.duplicated().sum()

np.int64(2939)

In [51]:
df_final[df_final.duplicated(keep=False)].head(20)

,Customer,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount,State,Gender
1068,GS98873,Bachelor,323912.47%,16061.0,88.0,1/0/00,Personal Auto,Four-Door Car,633.6,Arizona,F
1069,CW49887,Master,462680.11%,79487.0,114.0,1/0/00,Special Auto,SUV,547.2,California,F
1070,MY31220,College,899704.02%,54230.0,112.0,1/0/00,Personal Auto,Two-Door Car,537.6,California,F
1071,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1072,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1073,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1074,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1075,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1076,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1077,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [53]:
df_final = df_final.dropna(how="all")
df_final = df_final.reset_index(drop=True)

df_final.duplicated().sum()

np.int64(3)

In [54]:
df_final[df_final.duplicated(keep=False)]

,Customer,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount,State,Gender
1068,GS98873,Bachelor,323912.47%,16061.0,88.0,1/0/00,Personal Auto,Four-Door Car,633.6,Arizona,F
1069,CW49887,Master,462680.11%,79487.0,114.0,1/0/00,Special Auto,SUV,547.2,California,F
1070,MY31220,College,899704.02%,54230.0,112.0,1/0/00,Personal Auto,Two-Door Car,537.6,California,F
1071,GS98873,Bachelor,323912.47%,16061.0,88.0,1/0/00,Personal Auto,Four-Door Car,633.6,Arizona,F
1072,CW49887,Master,462680.11%,79487.0,114.0,1/0/00,Special Auto,SUV,547.2,California,F
1073,MY31220,College,899704.02%,54230.0,112.0,1/0/00,Personal Auto,Two-Door Car,537.6,California,F


In [55]:
df_final = df_final.drop_duplicates(keep="first")

In [56]:
df_final = df_final.reset_index(drop=True)

In [57]:
df_final.duplicated().sum()

np.int64(0)

In [58]:
df_final.shape

(9134, 11)

In [59]:
df_final.dtypes

Customer                         str
Education                        str
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                      str
Vehicle Class                    str
Total Claim Amount           float64
State                            str
Gender                           str
dtype: object

In [61]:
df_final["Customer Lifetime Value"] = (df_final["Customer Lifetime Value"].str.replace("%", ""))
df_final["Customer Lifetime Value"].head()

0           NaN
1     697953.59
2    1288743.17
3     764586.18
4     536307.65
Name: Customer Lifetime Value, dtype: object

In [62]:
df_final["Customer Lifetime Value"].dtype

dtype('O')

In [64]:
df_final["Customer Lifetime Value"] = pd.to_numeric(df_final["Customer Lifetime Value"])
df_final["Customer Lifetime Value"].dtype

dtype('float64')

In [65]:
df_final["Number of Open Complaints"].unique()

array(['1/0/00', '1/2/00', '1/1/00', '1/3/00', '1/5/00', '1/4/00', 0, 2,
       3, 1, 5, 4], dtype=object)

In [76]:
for x in df_final["Number of Open Complaints"].unique():
    if isinstance(x, str) and "/" not in x:
        print(repr(x))

'0'
'2'
'1'
'3'
'5'
'4'


In [77]:
df_final["Number of Open Complaints"] = df_final["Number of Open Complaints"].apply(
    lambda x: x.split("/")[1] if isinstance(x, str) and "/" in x else x
)

In [78]:
df_final["Number of Open Complaints"].unique()

array(['0', '2', '1', '3', '5', '4', 0, 2, 3, 1, 5, 4], dtype=object)

In [80]:
df_final["Number of Open Complaints"] = pd.to_numeric(
    df_final["Number of Open Complaints"]
)
df_final["Number of Open Complaints"].unique()

array([0, 2, 1, 3, 5, 4])

In [81]:
df_final.dtypes
df_final.isnull().sum()

Customer                        0
Education                       0
Customer Lifetime Value      7077
Income                          0
Monthly Premium Auto            0
Number of Open Complaints       0
Policy Type                     0
Vehicle Class                   0
Total Claim Amount              0
State                           0
Gender                        122
dtype: int64

In [82]:
df_final["Customer Lifetime Value"].count()
df_final["Customer Lifetime Value"].describe()
df_final[df_final["Customer Lifetime Value"].notna()].head()

,Customer,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount,State,Gender
1,QZ44356,Bachelor,697953.59,0.0,94.0,0,Personal Auto,Four-Door Car,1131.464935,Arizona,F
2,AI49188,Bachelor,1288743.17,48767.0,108.0,0,Personal Auto,Two-Door Car,566.472247,Nevada,F
3,WW63253,Bachelor,764586.18,0.0,106.0,0,Corporate Auto,SUV,529.881344,California,M
4,GA49547,High School or Below,536307.65,36357.0,68.0,0,Personal Auto,Four-Door Car,17.269323,Washington,M
5,OC83172,Bachelor,825629.78,62902.0,69.0,0,Personal Auto,Two-Door Car,159.383042,Oregon,F


In [83]:
df_final["Customer Lifetime Value"] = (
    df_final["Customer Lifetime Value"]
    .astype(str)
    .str.replace("%", "", regex=False)
)

df_final["Customer Lifetime Value"] = pd.to_numeric(
    df_final["Customer Lifetime Value"],
    errors="coerce"
)

In [84]:
df_final.isnull().sum()

Customer                        0
Education                       0
Customer Lifetime Value      7077
Income                          0
Monthly Premium Auto            0
Number of Open Complaints       0
Policy Type                     0
Vehicle Class                   0
Total Claim Amount              0
State                           0
Gender                        122
dtype: int64

In [85]:
df_new["Customer Lifetime Value"].isnull().sum()

np.int64(2944)

In [86]:
df_new["Customer Lifetime Value"].head(10)

0            NaN
1     697953.59%
2    1288743.17%
3     764586.18%
4     536307.65%
5     825629.78%
6     538089.86%
7     721610.03%
8    2412750.40%
9     738817.81%
Name: Customer Lifetime Value, dtype: object

In [87]:
df_clean = df_new.copy()

In [88]:
df_clean = df_clean.dropna(how="all")

In [89]:
df_clean.shape

(9137, 11)

In [90]:
df_clean["Customer Lifetime Value"].isnull().sum()

np.int64(7)

In [91]:
df_clean["Customer Lifetime Value"] = (
    df_clean["Customer Lifetime Value"]
    .astype(str)
    .str.replace("%", "", regex=False)
)

In [92]:
df_clean["Customer Lifetime Value"] = pd.to_numeric(
    df_clean["Customer Lifetime Value"],
    errors="coerce"
)

In [94]:
df_clean["Customer Lifetime Value"].dtype


dtype('float64')

In [95]:
df_clean["Customer Lifetime Value"].isnull().sum()

np.int64(7)

In [97]:
print("Mean:", df_clean["Customer Lifetime Value"].mean())
print("Median:", df_clean["Customer Lifetime Value"].median())

Mean: 182196.3896898556
Median: 7714.8768635


In [98]:
df_clean["Customer Lifetime Value"] = df_clean["Customer Lifetime Value"].fillna(
    df_clean["Customer Lifetime Value"].median()
)

In [99]:
df_clean["Customer Lifetime Value"].isnull().sum()

np.int64(0)

In [100]:
df_clean["Gender"].value_counts()

Gender
F    4607
M    4408
Name: count, dtype: int64

In [101]:
df_clean["Gender"].mode()

0    F
Name: Gender, dtype: str

In [102]:
df_clean["Gender"] = df_clean["Gender"].fillna(
    df_clean["Gender"].mode()[0]
)

In [103]:
df_clean.isnull().sum()

Customer                     0
Education                    0
Customer Lifetime Value      0
Income                       0
Monthly Premium Auto         0
Number of Open Complaints    0
Policy Type                  0
Vehicle Class                0
Total Claim Amount           0
State                        0
Gender                       0
dtype: int64

In [104]:
df_clean.duplicated().sum()

np.int64(3)

In [105]:
df_clean = df_clean.drop_duplicates(keep="first")

In [106]:
print(df_clean.shape)
print(df_clean.duplicated().sum())

(9134, 11)
0


In [107]:
df_clean.dtypes

Customer                         str
Education                        str
Customer Lifetime Value      float64
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                      str
Vehicle Class                    str
Total Claim Amount           float64
State                            str
Gender                           str
dtype: object

In [109]:
df_clean["Number of Open Complaints"] = df_clean["Number of Open Complaints"].apply(
    lambda x: x.split("/")[1] if isinstance(x, str) and "/" in x else x
)

In [110]:
df_clean["Number of Open Complaints"].unique()

array(['0', '2', '1', '3', '5', '4', 0, 2, 3, 1, 5, 4], dtype=object)

In [111]:
df_clean["Number of Open Complaints"] = pd.to_numeric(
    df_clean["Number of Open Complaints"]
)

In [112]:
df_clean["Number of Open Complaints"].dtype

dtype('int64')

In [113]:
numeric_columns = [
    "Customer Lifetime Value",
    "Income",
    "Monthly Premium Auto",
    "Total Claim Amount"
]

df_clean[numeric_columns] = df_clean[numeric_columns].astype(int)

In [114]:
df_clean.dtypes

Customer                       str
Education                      str
Customer Lifetime Value      int64
Income                       int64
Monthly Premium Auto         int64
Number of Open Complaints    int64
Policy Type                    str
Vehicle Class                  str
Total Claim Amount           int64
State                          str
Gender                         str
dtype: object

In [115]:
print("Shape:", df_clean.shape)
print("Null values:")
print(df_clean.isnull().sum())

print("\nDuplicates:", df_clean.duplicated().sum())

print("\nData types:")
print(df_clean.dtypes)

Shape: (9134, 11)
Null values:
Customer                     0
Education                    0
Customer Lifetime Value      0
Income                       0
Monthly Premium Auto         0
Number of Open Complaints    0
Policy Type                  0
Vehicle Class                0
Total Claim Amount           0
State                        0
Gender                       0
dtype: int64

Duplicates: 0

Data types:
Customer                       str
Education                      str
Customer Lifetime Value      int64
Income                       int64
Monthly Premium Auto         int64
Number of Open Complaints    int64
Policy Type                    str
Vehicle Class                  str
Total Claim Amount           int64
State                          str
Gender                         str
dtype: object


In [116]:
df_clean.head()

,Customer,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount,State,Gender
0,RB50392,Master,7714,0,1000,0,Personal Auto,Four-Door Car,2,Washington,F
1,QZ44356,Bachelor,697953,0,94,0,Personal Auto,Four-Door Car,1131,Arizona,F
2,AI49188,Bachelor,1288743,48767,108,0,Personal Auto,Two-Door Car,566,Nevada,F
3,WW63253,Bachelor,764586,0,106,0,Corporate Auto,SUV,529,California,M
4,GA49547,High School or Below,536307,36357,68,0,Personal Auto,Four-Door Car,17,Washington,M


In [117]:
df_clean.to_csv("cleaned_data.csv", index=False)

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [118]:
# Your code goes here
import pandas as pd

url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv"

df_marketing = pd.read_csv(url)

In [119]:
df_marketing.shape

(10910, 27)

In [120]:
df_marketing.head()

,unnamed:_0,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,...,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type,month
0,0,DK49336,Arizona,4809.216960,No,Basic,College,2011-02-18,Employed,M,...,9,Corporate Auto,Corporate L3,Offer3,Agent,292.800000,Four-Door Car,Medsize,A,2
1,1,KX64629,California,2228.525238,No,Basic,College,2011-01-18,Unemployed,F,...,1,Personal Auto,Personal L3,Offer4,Call Center,744.924331,Four-Door Car,Medsize,A,1
2,2,LZ68649,Washington,14947.917300,No,Basic,Bachelor,2011-02-10,Employed,M,...,2,Personal Auto,Personal L3,Offer3,Call Center,480.000000,SUV,Medsize,A,2
3,3,XL78013,Oregon,22332.439460,Yes,Extended,College,2011-01-11,Employed,M,...,2,Corporate Auto,Corporate L3,Offer2,Branch,484.013411,Four-Door Car,Medsize,A,1
4,4,QA50777,Oregon,9025.067525,No,Premium,Bachelor,2011-01-17,Medical Leave,F,...,7,Personal Auto,Personal L2,Offer1,Branch,707.925645,Four-Door Car,Medsize,A,1


In [121]:
df_marketing.dtypes

unnamed:_0                         int64
customer                             str
state                                str
customer_lifetime_value          float64
response                             str
coverage                             str
education                            str
effective_to_date                    str
employmentstatus                     str
gender                               str
income                             int64
location_code                        str
marital_status                       str
monthly_premium_auto               int64
months_since_last_claim          float64
months_since_policy_inception      int64
number_of_open_complaints        float64
number_of_policies                 int64
policy_type                          str
policy                               str
renew_offer_type                     str
sales_channel                        str
total_claim_amount               float64
vehicle_class                        str
vehicle_size    

In [123]:
df_marketing.isnull().sum()

unnamed:_0                       0
customer                         0
state                            0
customer_lifetime_value          0
response                         0
coverage                         0
education                        0
effective_to_date                0
employmentstatus                 0
gender                           0
income                           0
location_code                    0
marital_status                   0
monthly_premium_auto             0
months_since_last_claim          0
months_since_policy_inception    0
number_of_open_complaints        0
number_of_policies               0
policy_type                      0
policy                           0
renew_offer_type                 0
sales_channel                    0
total_claim_amount               0
vehicle_class                    0
vehicle_size                     0
vehicle_type                     0
month                            0
dtype: int64

In [124]:
df_marketing.duplicated().sum()

np.int64(0)

In [125]:
df_marketing["unnamed:_0"].nunique()

10910

In [126]:
df_marketing = df_marketing.drop(columns=["unnamed:_0"])

In [127]:
df_marketing.shape

(10910, 26)

In [128]:
df_marketing.duplicated().sum()

np.int64(443)

In [129]:
df_marketing[df_marketing.duplicated(keep=False)].sort_values("customer").head(20)

,customer,state,customer_lifetime_value,response,coverage,education,effective_to_date,employmentstatus,gender,income,...,number_of_policies,policy_type,policy,renew_offer_type,sales_channel,total_claim_amount,vehicle_class,vehicle_size,vehicle_type,month
3960,AA10041,California,7901.744238,No,Basic,High School or Below,2011-02-02,Unemployed,M,0,...,5,Personal Auto,Personal L3,Offer1,Call Center,1258.327804,SUV,Medsize,A,2
8731,AA10041,California,7901.744238,No,Basic,High School or Below,2011-02-02,Unemployed,M,0,...,5,Personal Auto,Personal L3,Offer1,Call Center,1258.327804,SUV,Medsize,A,2
7527,AE74442,California,7379.456823,No,Basic,Bachelor,2011-01-21,Employed,F,61584,...,2,Corporate Auto,Corporate L2,Offer1,Agent,205.294927,Four-Door Car,Medsize,A,1
6792,AE74442,California,7379.456823,No,Basic,Bachelor,2011-01-21,Employed,F,61584,...,2,Corporate Auto,Corporate L2,Offer1,Agent,205.294927,Four-Door Car,Medsize,A,1
802,AI49712,Washington,2593.897887,No,Basic,High School or Below,2011-01-21,Unemployed,F,0,...,1,Personal Auto,Personal L1,Offer3,Call Center,345.600000,Four-Door Car,Large,A,1
8719,AI49712,Washington,2593.897887,No,Basic,High School or Below,2011-01-21,Unemployed,F,0,...,1,Personal Auto,Personal L1,Offer3,Call Center,345.600000,Four-Door Car,Large,A,1
3500,AK59340,California,5601.138061,No,Basic,High School or Below,2011-01-05,Employed,M,42360,...,8,Corporate Auto,Corporate L3,Offer1,Branch,99.509191,Four-Door Car,Medsize,A,1
3577,AK59340,California,5601.138061,No,Basic,High School or Below,2011-01-05,Employed,M,42360,...,8,Corporate Auto,Corporate L3,Offer1,Branch,99.509191,Four-Door Car,Medsize,A,1
6015,AK59340,California,5601.138061,No,Basic,High School or Below,2011-01-05,Employed,M,42360,...,8,Corporate Auto,Corporate L3,Offer1,Branch,99.509191,Four-Door Car,Medsize,A,1
8474,AK59340,California,5601.138061,No,Basic,High School or Below,2011-01-05,Employed,M,42360,...,8,Corporate Auto,Corporate L3,Offer1,Branch,99.509191,Four-Door Car,Medsize,A,1


In [130]:
df_marketing[df_marketing.duplicated(keep=False)].shape

(858, 26)

In [131]:
df_marketing = df_marketing.drop_duplicates(keep="first")

In [132]:
df_marketing = df_marketing.reset_index(drop=True)

In [133]:
print(df_marketing.shape)
print(df_marketing.duplicated().sum())

(10467, 26)
0


In [134]:
df_marketing["effective_to_date"] = pd.to_datetime(
    df_marketing["effective_to_date"]
)

In [135]:
df_marketing["effective_to_date"].dtype

dtype('<M8[us]')

In [136]:
df_marketing[["effective_to_date", "month"]].head(10)


,effective_to_date,month
0,2011-02-18,2
1,2011-01-18,1
2,2011-02-10,2
3,2011-01-11,1
4,2011-01-17,1
5,2011-02-14,2
6,2011-02-14,2
7,2011-02-24,2
8,2011-01-19,1
9,2011-01-04,1


In [137]:
df_marketing["education"].value_counts()

education
Bachelor                3127
College                 3059
High School or Below    3026
Master                   866
Doctor                   389
Name: count, dtype: int64

In [138]:
df_marketing["vehicle_class"].value_counts()

vehicle_class
Four-Door Car    5498
Two-Door Car     2074
SUV              1972
Sports Car        537
Luxury SUV        203
Luxury Car        183
Name: count, dtype: int64

In [ ]:
##Question 1

# Create a pivot table with total revenue by sales channel

revenue_by_channel = df_marketing.pivot_table(
    index="sales_channel",
    values="total_claim_amount",
    aggfunc="sum"
)

# Round to 2 decimal places
revenue_by_channel = revenue_by_channel.round(2)

# Display the result
print(revenue_by_channel)

               total_claim_amount
sales_channel                    
Agent                  1748923.49
Branch                 1259333.44
Call Center             889094.77
Web                     676658.10


In [ ]:
### Analysis

The pivot table shows the total claim amount for each sales channel.
The sales channel with the highest total value represents the largest
contribution to the total amount in the dataset.

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [140]:
## Question 2

# Create a pivot table with average Customer Lifetime Value
# by gender and education level

clv_gender_education = df_marketing.pivot_table(
    index="gender",
    columns="education",
    values="customer_lifetime_value",
    aggfunc="mean"
)

# Round to 2 decimal places
clv_gender_education = clv_gender_education.round(2)

# Display the result
print(clv_gender_education)

education  Bachelor  College   Doctor  High School or Below   Master
gender                                                              
F           7913.45  7755.03  7369.59               8616.66  8226.36
M           7713.30  8086.29  7414.86               8162.36  8154.61


In [ ]:
### Analysis

The pivot table compares the average Customer Lifetime Value across 
gender and education levels. The results show differences in average
CLV between the different customer segments.

### Analysis Question 1 — Total Revenue by Sales Channel

The Agent sales channel has the highest total revenue, with 1,748,923.49,
followed by Branch with 1,259,333.44.

Call Center generated 889,094.77, while Web had the lowest total revenue
at 676,658.10.

Based on the results, Agent is the leading sales channel in terms of
total revenue.

### Analysis Question 2 — Average Customer Lifetime Value by Gender and Education

The average Customer Lifetime Value varies across gender and education levels.

Female customers with a High School education or below have the highest
average CLV at 8,616.66. For male customers, the highest average CLV is
also found in the High School or Below group, at 8,162.36.

Customers with a Doctor education level have the lowest average CLV for
both genders.

There is no consistent pattern showing that one gender has a higher CLV
across all education levels. This suggests that both gender and education
should be considered when analyzing customer value.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [3]:
# Your code goes here